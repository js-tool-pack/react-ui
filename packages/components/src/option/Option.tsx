import React from 'react';
import type { OptionProps } from './option.types';
import { getComponentClass, getSizeClassName } from '@pkg/shared';
import type { RequiredPart } from '@tool-pack/types';
import { getClassNames } from '@tool-pack/basic';
import { Icon } from '~/icon';

const defaultProps = {
  tag: 'div',
  size: 'medium',
  attrs: { role: 'option' },
} satisfies Partial<OptionProps>;

const rootName = getComponentClass('option');

export const Option: React.FC<OptionProps> = React.forwardRef<
  HTMLElement,
  OptionProps
>((props, ref) => {
  const { children, icon, readonly, extra, size, tag, disabled } =
    props as RequiredPart<OptionProps, keyof typeof defaultProps>;

  const attrs = {
    ...defaultProps.attrs,
    ...props.attrs,
  } as OptionProps['attrs'];

  const Child = (
    <>
      {icon && <Icon className={`${rootName}__icon`}>{icon}</Icon>}
      <div className={`${rootName}__label`}>{children}</div>
      {extra && <div className={`${rootName}__extra`}>{extra}</div>}
    </>
  );

  return React.createElement(
    tag,
    {
      ...attrs,
      ref,
      disabled,
      className: getClassNames(
        rootName,
        attrs?.className,
        getSizeClassName(size),
        {
          [`${rootName}--readonly`]: readonly,
        },
      ),
    },
    Child,
  );
});

Option.defaultProps = defaultProps;
Option.displayName = 'Option';
