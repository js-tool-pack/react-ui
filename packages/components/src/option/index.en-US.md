---
category: Components
title: Option
atomId: Option
debug: true
demo:
  cols: 2
group:
  title: Inner
---

Option 说明。

## 代码演示

<!-- prettier-ignore -->
<code src="./demo/basic.tsx"></code>
<code src="./demo/readonly.tsx"></code>
<code src="./demo/disabled.tsx"></code>
<code src="./demo/icon.tsx"></code>
<code src="./demo/extra.tsx"></code>

## API

Option 的属性说明如下：

| 属性     | 说明     | 类型            | 默认值 | 版本 |
| -------- | -------- | --------------- | ------ | ---- |
| readonly | 只读     | boolean         | --     | --   |
| disabled | 禁用     | boolean         | --     | --   |
| icon     | 图标     | React.ReactNode | --     | --   |
| extra    | 额外元素 | React.ReactNode | --     | --   |

其他说明。
