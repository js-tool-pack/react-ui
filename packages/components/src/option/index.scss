@use '../namespace' as Name;

$r: Name.$option;
:root {
  --t-option-active-bg-color: rgb(243 243 245);
  &.dark {
    --t-option-active-bg-color: rgb(255 255 255 / 9%);
  }
}
.#{$r} {
  --t-option-size: var(--t-size);
  --t-option-padding: var(--t-radius);
  --t-option-color: var(--t-text-color);
  --t-option-active-color: white;
  --t-option-disabled-color: var(--t-disabled-text-color);
  --t-option-bg-color: var(--t-bg-color);
  --t-option-border-radius: var(--t-radius);
  --t-option-transition-duration: var(--t-transition-duration);

  position: relative;
  display: flex;
  align-items: center;
  padding: 0 var(--t-option-padding);
  min-height: var(--t-option-size);
  border-radius: var(--t-option-border-radius);
  color: var(--t-option-color);
  background-color: var(--t-option-bg-color);

  // fixme: "transition-property: all;"会导致 Transition 组件 hide 后闪烁
  // transition: all var(--t-option-transition-duration);

  transition-property: background-color, color;
  transition-duration: var(--t-option-transition-duration);
  &--readonly {
    --t-option-color: var(--t-option-disabled-color);
  }
  &[disabled] {
    --t-option-color: var(--t-option-disabled-color);

    cursor: not-allowed;
  }
  &:not([disabled]):not(.#{$r}--readonly) {
    cursor: pointer;
    &:hover {
      --t-option-bg-color: var(--t-option-active-bg-color);
    }
  }
  &--active {
    background-color: var(--t-primary-bg-color);
  }
  &__label {
    flex: 1;
  }
  &__icon {
    margin-right: 6px;
  }
  &__extra {
    display: flex;
    align-items: center;
    margin-left: 1em;
  }
}
