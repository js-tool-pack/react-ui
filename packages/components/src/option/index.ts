export type { OptionProps } from './option.types';
export * from './Option';
