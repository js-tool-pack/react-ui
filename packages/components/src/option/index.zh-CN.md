---
category: Components
title: Option 选项
atomId: Option
debug: true
demo:
  cols: 2
group:
  title: 通用
---

Option 说明。

## 代码演示

<!-- prettier-ignore -->
<code src="./demo/basic.tsx"></code>
<code src="./demo/readonly.tsx"></code>
<code src="./demo/disabled.tsx"></code>
<code src="./demo/icon.tsx"></code>
<code src="./demo/expandable.tsx"></code>

## API

Option 的属性说明如下：

| 属性 | 说明 | 类型 | 默认值 | 版本 |
| ---- | ---- | ---- | ------ | ---- |
| --   | --   | --   | --     | --   |

其他说明。
