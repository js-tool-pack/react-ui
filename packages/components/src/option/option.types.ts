import React from 'react';
import type { PropsBase, Size } from '@pkg/shared';

export interface OptionProps extends PropsBase {
  tag?: keyof HTMLElementTagNameMap;
  size?: Size;
  disabled?: boolean;
  expandable?: boolean;
  readonly?: boolean;
  icon?: React.ReactNode;
}
