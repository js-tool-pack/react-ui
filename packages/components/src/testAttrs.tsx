import { fireEvent, render } from '@testing-library/react';
import type { PropsBase } from '@pkg/shared';
import React from 'react';

export function testAttrs<
  T extends { attrs?: Partial<React.HTMLAttributes<HTMLElement>> },
>(component: React.FC<T>, displayName?: string) {
  const name = displayName || component.displayName || component.name;
  test(name ? `${name} attrs` : `attrs`, () => {
    const onClick = jest.fn();

    const FC = component as React.FC<PropsBase>;
    const { container } = render(
      <FC
        attrs={{
          style: { width: '20px' },
          className: 'foo',
          role: 'note',
          onClick,
        }}
      />,
    );

    expect(container.firstChild).toHaveClass('foo');
    expect(container.firstChild).toHaveAttribute('role', 'note');
    expect(container.firstChild).toHaveStyle({ width: '20px' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(container.firstChild!);
    expect(onClick).toBeCalled();
  });
}
