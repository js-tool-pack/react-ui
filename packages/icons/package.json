{
  "version": "0.0.2-beta.3",
  "types": "dist/index.d.ts",
  "main": "dist/index.js",
  "private": true,
  "scripts": {
    "rm:dist": "rimraf ./dist",
    "build:this": "tsc -p tsconfig.json",
    "build": "pnpm rm:dist && pnpm build:this",
    "check:ts": "tsc --noEmit -p tsconfig.json"
  },
  "devDependencies": {}
}
