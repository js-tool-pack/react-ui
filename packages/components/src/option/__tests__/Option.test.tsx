import { testAttrs } from '~/testAttrs';
import { Option } from '~/option';
import { render } from '@testing-library/react';
import { CircleClose } from '@pkg/icons';

describe('Option', () => {
  testAttrs(Option);

  test('snap', () => {
    const { container } = render(<Option>foo bar</Option>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('disabled', () => {
    const { container } = render(<Option disabled>foo bar</Option>);
    expect(container.firstChild).toHaveAttribute('disabled', '');
  });

  test('extra', () => {
    const { container } = render(
      <Option extra={'hello world'}>foo bar</Option>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('icon', () => {
    const { container } = render(
      <Option icon={<CircleClose />}>foo bar</Option>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('readonly', () => {
    const { container } = render(<Option readonly>foo bar</Option>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
