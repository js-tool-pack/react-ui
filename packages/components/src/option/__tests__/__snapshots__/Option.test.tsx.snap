// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option extra 1`] = `
<div
  class="t-option t--size-m"
  role="option"
>
  <div
    class="t-option__label"
  >
    foo bar
  </div>
  <div
    class="t-option__extra"
  >
    hello world
  </div>
</div>
`;

exports[`Option icon 1`] = `
<div
  class="t-option t--size-m"
  role="option"
>
  <i
    class="t-icon t-option__icon"
    role="img"
  >
    <svg
      viewBox="0 0 1024 1024"
    >
      <path
        d="M512 14.8973037C238.59351703 14.8973037 14.8973037 238.59351703 14.8973037 512s223.69621333 497.1026963 497.1026963 497.1026963 497.1026963-223.69621333 497.1026963-497.1026963S785.40648297 14.8973037 512 14.8973037zM512 926.25224691c-231.98125827 0-414.25224691-182.27098864-414.25224691-414.25224691s182.27098864-414.25224691 414.25224691-414.25224691 414.25224691 182.27098864 414.25224691 414.25224691S743.98125827 926.25224691 512 926.25224691z"
        fill="currentColor"
      />
      <path
        d="M719.12612347 304.87387653c-16.57008987-16.57008987-41.42522469-16.57008987-57.99531458 0L512 454.00468543 362.86919111 304.87387653C346.29910124 288.30378667 321.44396642 288.30378667 304.87387653 304.87387653S288.30378667 346.29910124 304.87387653 362.86919111L454.00468543 512 304.87387653 661.13080889c-16.57008987 16.57008987-16.57008987 41.42522469 0 57.99531458s41.42522469 16.57008987 57.99531458 0L512 569.99531457l149.13080889 149.1308089c16.57008987 16.57008987 41.42522469 16.57008987 57.99531458 0s16.57008987-41.42522469 0-57.99531458L569.99531457 512l149.1308089-149.13080889C735.69621333 346.29910124 735.69621333 321.44396642 719.12612347 304.87387653z"
        fill="currentColor"
      />
    </svg>
  </i>
  <div
    class="t-option__label"
  >
    foo bar
  </div>
</div>
`;

exports[`Option readonly 1`] = `
<div
  class="t-option t--size-m t-option--readonly"
  role="option"
>
  <div
    class="t-option__label"
  >
    foo bar
  </div>
</div>
`;

exports[`Option snap 1`] = `
<div
  class="t-option t--size-m"
  role="option"
>
  <div
    class="t-option__label"
  >
    foo bar
  </div>
</div>
`;
